textbook/2_grouping.qmd

```diff
@@ -0,0 +1,994 @@
+
+
+```{python}
+alt.Chart(income_df).mark_point().encode(
+    x='life_expectancy_at_50',
+    y='income_group',
+    color='region',
+)
+```
+
+:::
+
+# ggplot
+
+```{webr fig.width=8, fig.height=6}
+#| setup: true
+#| exercise:
+#|   - position_ggplot
+#|   - facet_ggplot
+library(ggplot2)
+
+url <- 'data/income_lifeexp_alcohol_1.csv'
+```
+
+```{webr fig.width=8, fig.height=6}
+#| exercise: position_ggplot
+income_df <- readr::read_csv(url, show_col_types=FALSE)
+
+
+
+
+
+# Write code above to create the chart 
+```
+
+::: {.callout-default .ex-solution}
+
+You are primarily comparing the `income_group` variable, so it should be grouped by position on either the `x` or `y` axis. The secondary comparison, `region`, should be encoded as color. 
+```{r fig.width=8, fig.height=6}
+ggplot(income_df, aes(
+    x=life_expectancy_at_50, 
+    y=income_group, 
+    color=region
+)) + 
+geom_point()
+```
+
+:::
+
+:::
+
+:::::
+
+For all of these benefits, there are some drawbacks of grouping by position. For one, it occupies an axis that could have represented a continuous variable. This reduces the information density of your visualization. More worryingly, visually separating each group by position can hide overall trends in the dataset. It also requires significant space.
+
+Lastly, plotting data points along a single axis can lead to  overlap. That's where **jittering** comes to the rescue. Jittering introduces slight variation, or 'jitter', in the position of data points in each lane, making them easier to see. 
+
+::: {.column-margin}
+
+**Jitterbug**
+
+Back in the 1930s, the jitterbug was a lively, often wild, dance style that matched the fast-paced, unpredictable energy of swing music. In data visualization, jittering serves a similar purpose: to add a touch of randomness to points in a plot.
+
+![](/img/640px-Jitterbug_Wolcott_FSA.jpg)
+
+:::
+
+Let's examine how jittering works.
+
+::: {.panel-tabset}
+
+# Altair
+
+In Altair, we create a random calculated variable called `jitter` by appending the `.transform_calculate()` method to our chart. We use our newly created `jitter` variable, encoding it as the `yOffset` of each data point using `yOffset='jitter:Q'`. This shifts the `yOffset` of each point by a random value. 
+
+The `:Q` simply tells Altair that `jitter` is a continuous quantitative variable.
+
+```{python}
+alt.Chart(income_df).mark_point().encode(
+    x='percentage_heavy_drinkers',
+    y='region',
+    color='region',
+    yOffset="jitter:Q"
+).transform_calculate(
+    jitter="random()"
+)
+```
+
+# ggplot
+
+In ggplot2, you can achieve the same effect with jittering by using the `position_jitter()` function. Jitter is directly applied to the marks through the position argument within `geom_point()`.
+
+```{r fig.width=8, fig.height=6}
+ggplot(income_df, aes(
+    x=percentage_heavy_drinkers, 
+    y=region, 
+    color=region
+)) + 
+geom_point(position = position_jitter(width = 0, height = 0.2))
```

Comment:
```suggestion
In ggplot2, you can jitter points by using the `geom_jitter`, which is a combination of `geom_point` together with a function called `position_jitter`.

```{r fig.width=8, fig.height=6}
ggplot(income_df, aes(
    x=percentage_heavy_drinkers, 
    y=region, 
    color=region
)) + 
geom_jitter()
```

Was there a specific reason you opted for position_jitter here? Maybe because it is more explicit? I see geom_jitter more commonly, but happy to switch if you think the other approach is more instructive.